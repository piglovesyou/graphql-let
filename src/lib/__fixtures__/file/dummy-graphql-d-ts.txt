/* a7870383395320c75dc806f677ffa7caad7522bc
 * This file is automatically generated by graphql-let.
 */
