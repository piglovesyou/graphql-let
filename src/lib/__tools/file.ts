import copyfiles from 'copyfiles';
import { promises } from 'fs';
import { join as pathJoin, sep } from 'path';
import _rimraf from 'rimraf';
import { promisify } from 'util';

export const rimraf = promisify(_rimraf);

const crlf = /\r\n/g;
export function normalizeNewLine(str: string) {
  if (crlf.test(str)) return str.replace(crlf, '\n');
  return str;
}

export const { readFile: _readFile, writeFile, rename } = promises;

export function readFile(file: string) {
  return _readFile(file, 'utf-8').then(normalizeNewLine);
}

export function cleanup(cwd: string, relPaths: string[]) {
  return Promise.all(relPaths.map((rel) => rimraf(pathJoin(cwd, rel))));
}

export function copyDir(
  baseFullDir: string,
  srcRelDir: string,
  destRelDir: string,
): Promise<void> {
  // baseFullDir: /Users/a/b/c
  // srcRelDir: d/e
  // destRelDir: .d/e
  // It generates /Users/a/b/c/.d/e
  if (srcRelDir === destRelDir) throw new Error('Kidding me?');
  const up = pathJoin(baseFullDir).split(sep).length + 1;
  const [relRoot] = destRelDir.split('/');
  return new Promise<void>((resolve, rejects) => {
    copyfiles(
      [pathJoin(baseFullDir, srcRelDir, '**'), pathJoin(baseFullDir, relRoot)],
      { error: true, up, all: true },
      (err) => {
        if (err) return rejects(err);
        resolve();
      },
    );
  });
}

export type AbsFn = (rel: string) => string;

export async function prepareFixtures(
  baseFullDir: string,
  fixtureSrcRelDir: string,
): Promise<[cwd: string, abs: AbsFn]> {
  if (fixtureSrcRelDir.startsWith('..'))
    throw new Error(
      `It doesn't support backward relative paths like ${fixtureSrcRelDir}`,
    );
  const fixtureDestRelDir = '.' + fixtureSrcRelDir;
  const cwd = pathJoin(baseFullDir, fixtureDestRelDir);
  const abs = (relPath: string) => pathJoin(cwd, relPath);
  await copyDir(baseFullDir, fixtureSrcRelDir, fixtureDestRelDir);
  return [cwd, abs];
}
