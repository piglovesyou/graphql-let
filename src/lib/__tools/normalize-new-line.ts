const crlf = /\r\n/g;

export function normalizeNewLine(str: string) {
  if (crlf.test(str)) return str.replace(crlf, '\n');
  return str;
}
