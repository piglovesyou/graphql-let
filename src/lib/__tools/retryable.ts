function timeout(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export default async function retryable(
  fn: () => Promise<void>,
  interval: number,
  maxTimeout: number,
) {
  if (maxTimeout < 0) throw new Error('never');
  const end = Date.now() + maxTimeout;
  let lastError: Error = new Error('never');
  let n = 0;
  while (Date.now() < end) {
    try {
      await fn();
      console.info(`Succeeded on the ${++n} time.`);
      return;
    } catch (e) {
      console.info(`${++n} times tried. Retrying...`);
      lastError = e;
      await timeout(interval);
    }
  }
  throw lastError;
}
