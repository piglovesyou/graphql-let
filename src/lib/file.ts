import { promises as fsPromises } from 'fs';
import { createReadStream, existsSync } from 'fs';
import glob from 'globby';
import _rimraf from 'rimraf';
import { promisify } from 'util';

export const rimraf = promisify(_rimraf);

export const { readFile, writeFile } = fsPromises;
export { readFileSync, statSync } from 'fs';

// Erasing old cache in __generated__ on HMR.
// Otherwise the multiple `declare module "*/x.graphql"` are exposed.
export async function removeByPatterns(cwd: string, ...patterns: string[]) {
  const oldFiles = await glob(patterns, {
    cwd,
    absolute: true,
  });
  await Promise.all(oldFiles.map((f) => rimraf(f)));
}

const leadingStringOfGeneratedContent = '/* ';
const hexHashLength = 40;
export function withHash(sourceHash: string, content: string): string {
  return `${leadingStringOfGeneratedContent}${sourceHash}
 * This file is automatically generated by graphql-let. */

${content}`;
}

export function readHash(filePath: string): Promise<string | null> {
  if (!existsSync(filePath)) {
    return Promise.resolve(null);
  }

  return new Promise((resolve, reject) => {
    const stream = createReadStream(filePath, {
      encoding: 'utf-8',
      highWaterMark: leadingStringOfGeneratedContent.length + hexHashLength,
    });
    stream.on('error', (error) => reject(error));
    stream.on('data', (chunk: string) => {
      const hash = chunk.slice(leadingStringOfGeneratedContent.length);
      if (hash.length !== hexHashLength) return resolve(null);

      resolve(hash);
    });
    stream.read();
  });
}
