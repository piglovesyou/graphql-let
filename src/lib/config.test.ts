import loadConfig, { loadConfigSync } from './config';
import { prepareFixtures } from './__tools/file';

let cwd: string;

describe('config.ts', () => {
  beforeAll(async () => {
    [cwd] = await prepareFixtures(__dirname, '__fixtures/config');
    process.env.GRAPHQL_SERVER_ENDPOINT = 'https://yeah/graphql';
    process.env.GRAPHQL_SERVER_TOKEN = 'blaa';
  });

  afterAll(() => {
    delete process.env.GRAPHQL_SERVER_ENDPOINT;
    delete process.env.GRAPHQL_SERVER_TOKEN;
  });

  test('loads config with default values', async () => {
    const actual = await loadConfig(cwd, '.graphql-let-simple.yml');
    expect(actual).toMatchSnapshot();
  });

  test('interpolates environment variables', () => {
    const [{ schema }] = loadConfigSync(cwd, '.graphql-let-envvar.yml');
    expect(schema).toMatchSnapshot();
  });

  test('interpolates environment variables asynchronously', async () => {
    const [{ schema }] = await loadConfig(cwd, '.graphql-let-envvar.yml');
    expect(schema).toMatchSnapshot();
  });

  test('overwrite default values', async () => {
    const [actual] = await loadConfig(cwd, '.graphql-let-overwrite.yml');
    expect(actual).toMatchSnapshot();
  });
});
