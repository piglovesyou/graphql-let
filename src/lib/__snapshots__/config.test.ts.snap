// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config.ts interpolates environment variables 1`] = `
Array [
  Object {
    "https://yeah/graphql": Object {
      "headers": Object {
        "Authorization": "blaa",
      },
    },
  },
]
`;

exports[`config.ts interpolates environment variables asynchronously 1`] = `
Array [
  Object {
    "https://yeah/graphql": Object {
      "headers": Object {
        "Authorization": "blaa",
      },
    },
  },
]
`;

exports[`config.ts loads config with default values 1`] = `
Array [
  Object {
    "TSConfigFile": "tsconfig.json",
    "cacheDir": "node_modules/graphql-let/__generated__",
    "documents": Array [
      "**/*.graphql",
    ],
    "generateOptions": Object {},
    "gqlDtsEntrypoint": "node_modules/@types/graphql-let/index.d.ts",
    "plugins": Array [
      "typescript",
    ],
    "respectGitIgnore": true,
    "schema": Array [
      Object {
        "https://yeah/graphql": Object {
          "headers": Object {
            "Authorization": "blaa",
          },
        },
      },
    ],
    "schemaEntrypoint": "",
  },
  "6134bbfa2c25f64acb7c526097ecc4e4014c2aa4",
]
`;

exports[`config.ts overwrite default values 1`] = `
Object {
  "TSConfigFile": "tsconfig.yeah.json;",
  "cacheDir": "__yeah__;",
  "documents": Array [
    "**/*.graphql",
  ],
  "generateOptions": Object {},
  "gqlDtsEntrypoint": "typings/yeah.d.ts;",
  "plugins": Array [
    "typescript",
  ],
  "respectGitIgnore": false,
  "schema": "**/*.graphql",
  "schemaEntrypoint": "yeah.graphqls",
}
`;
